fingerprint/infra/nec-bio-sdk/src/main/java/com/simprints/fingerprint/infra/necsdkimpl/matching/FingerprintMatcherImpl.kt

```diff
@@ -0,0 +1,118 @@
+        return if (settings?.crossFingerComparison == true) {
+            crossFingerMatching(probe, candidates)
+        } else {
+            sameFingerMatching(probe, candidates)
+        }
+    }
+
+    private fun sameFingerMatching(
+        probe: FingerprintIdentity,
+        candidates: List<FingerprintIdentity>
+    ) = candidates.map {
+        sameFingerMatching(probe, it)
+    }
+
+    /**
+     * This method gets the matching score
+     * the matching score is calculated by
+     * - Getting the matching score for each similar finger pairs using NEC SDK
+     * - The overall score is the average of the individual finger match scores
+     * - We ignore probe fingers that doesn't have matching candidate fingers
+     * @param probe
+     * @param candidate
+     * @return MatchResult
+     */
+    private fun sameFingerMatching(
+        probe: FingerprintIdentity,
+        candidate: FingerprintIdentity
+    ): MatchResult {
+        var fingers = 0 // the number of fingers used in matching
+        val total = probe.fingerprints.sumOf { fingerprint ->
+            // we should ignore probe fingers that doesn't have matching candidate fingers
+
+            require(fingerprint.format == NEC_TEMPLATE_FORMAT)
+            candidate.templateForFinger(fingerprint.fingerId)?.let { candidateTemplate ->
+                require(candidateTemplate.format == NEC_TEMPLATE_FORMAT)
+                fingers++
+                verify(fingerprint, candidateTemplate)
+            } ?: 0.toDouble()
+        }
+        return MatchResult(candidate.id, getOverallScore(total, fingers))
+    }
+
+    private fun verify(probe: Fingerprint, candidate: Fingerprint) = try {
+        nec.match(
+            probe.toNecTemplate(),
+            candidate.toNecTemplate()
+        ).toDouble()
+    } catch (e: Exception) {
+        throw BioSdkException.TemplateMatchingException(e)
+    }
+
+    private fun crossFingerMatching(
+        probe: FingerprintIdentity,
+        candidates: List<FingerprintIdentity>
+    ) = candidates.map { crossFingerMatching(probe, it) }
+
+    private fun crossFingerMatching(
+        probe: FingerprintIdentity,
+        candidate: FingerprintIdentity,
+    ): MatchResult {
+        // Number of fingers used in matching
+        val fingers = probe.fingerprints.size
+        // Sum of maximum matching score for each finger
+        val total = probe.fingerprints
+            .sumOf { probeTemplate ->
+                candidate.fingerprints
+                    .maxOf { candidateTemplate ->
+                        verify(
+                            probeTemplate,
+                            candidateTemplate
+                        )
+                    }
+            }
+        // Matching score  = total/number of fingers
+        return MatchResult(candidate.id, getOverallScore(total, fingers))
+    }
+
+}
+
+private fun FingerprintIdentity.templateForFinger(fingerId: FingerIdentifier) = fingerprints.find {
```

Comment:
This is a matter of preference, but I do not see a reason for these functions to be at top level.